non blocking nRF loop

rejoe2 opened the pull request:
Seems to work very good for 1/2channel models
Still has problems with (some!) 4ch samples

rejoe2 committed:
non blocking version

- shorten hms waiting times after last frame es well (might be too short to immedately stop timer!)
- timings still need review...

rejoe2 committed:
Disable stopping timer for CMT 

breaks communication when enabled. Maybe there's another solution or some bug?

stefan123t commented on src/hm/hmRadio.h:
```diff
@@ -105,61 +105,88 @@ class HmRadio : public Radio {
+            if(!mIrqRcvd) {     // no news from nRF, check timers
+                if (mRadioWaitTime.isTimeout()) { // timeout reached!
+                    mNRFisInRX = false;
+                    // add stop listening?
```
Eigentlich könnten wir den NRF ja im default / immer im RX mode laufen lassen, das ist dann schon fast wie der Promiscuous Mode.
Nur wenn wir eine Anfrage senden wollen, müssen wir in den TX mode wechseln.

rejoe2 replied:
Weiß nicht: Die eingehenden Nachrichten müssen ja irgendwie einem Inverter zugeordnet werden, und das macht der rx-Teil im Moment nicht (mehr? war in den alten Fassungen evtl. anders). 
Und auch die Kanalwechsel müßte man organisieren. Nach welcher Vorgabe?

stefan123t commented on src/hm/Communication.h:
```diff
@@ -321,6 +309,56 @@ class Communication : public CommQueue<> {
             }
         }
 
+
```
Eventuell jeweils ein Array `InverterRealTimePackets[]={7, 2, 3, 4, 7}` mit einem 
`INV_TYPE enum(INV_TYPE_DEFAULT=0, INV_TYPE_1CH, INV_TYPE_2CH, INV_TYPE_4CH, INV_TYPE_6CH)` als Index ?
Dann spart man sich die vielen switch-case bzw. if/then/else Bedingungen.

rejoe2 replied:
Eleganter Vorschlag!
War erst mal beschäftigt, überhaupt die passende Logik zusammenzufrickeln...

stefan123t commented on src/hm/Communication.h:
```diff
@@ -787,9 +823,11 @@ class Communication : public CommQueue<> {
             q->iv->radio->sendCmdPacket(q->iv, cmd, 0x00, true);
 
-            mWaitTime.startTimeMonitor(MI_TIMEOUT);
+            q->iv->radio->mRadioWaitTime.startTimeMonitor(DURATION_TXFRAME + DURATION_ONEFRAME + DURATION_RESERVE);
```
Das taucht noch dreimal auf, aber das müsste der Optimizer zusammenrechnen sind ja alles konstante Defines ?

rejoe2 replied:
Der Ansatz war erst mal, eine funktionierende Lösung zu finden. 
Vermutlich ist da noch Nacharbeit erforderlich, die timings sind uU. für CMT-Inverter im Moment geringfügig zu kurz (der erste frame kommt da anscheinend per default sehr viel später, was sich aber nur bei sehr langen payloads auswirkt?)

stefan123t commented on src/hm/hmInverter.h:
```diff
@@ -849,7 +855,7 @@ class Inverter {
         uint16_t mIvRxCnt  = 0;
         uint16_t mIvTxCnt  = 0;
-
+        uint16_t mAckCount = 0;
```
Cool, gibt es die auch in der Statistik dann ?

rejoe2 replied:
Fände ich cool, würde dann aber gleich mit anregen, die "temporären Daten" (loss rates) dann auch optisch etwas abzusetzen von den Gesamt-Statistik-Daten.

Es gibt aber leider nur sehr wenige User, die sowas überhaupt gesehen haben...

knickohr replied:
Sollte dieser Ack Count schon in der 5801 sichtbar sein ?

stefan123t commented on src/hm/hmRadio.h:
```diff
@@ -105,61 +105,88 @@ class HmRadio : public Radio {
+                innerLoopTimeout = DURATION_TXFRAME;
+            }
+
+            if(rx_ready) {
```
Das würde ich nach oben schieben, das ist m.E. das Timing kritischere und der default ist Empfangen (RX),

rejoe2 replied:
Kann man machen, aber in meiner idealen Welt gibt es sowieso nur einen der drei Anlässe, nie mehrere gleichzeitig. Von daher war mein Angang für den unwahrscheinlichen Fall, dass doch alles zusammenkommt gewesen, zuerst den wichtigen Teil zu erledigen, und das ist das Zurückschalten auf rx nach tx.

stefan123t commented on src/hm/hmRadio.h:
```diff
@@ -285,16 +311,18 @@ class HmRadio : public Radio {
                                 ah::dumpBuf(p.packet, p.len, 1, 4);
```
Sollten wir hier nicht erst prüfen ob es ein TX Command (MainCmd bzw `p.packet[0] < 0x80`) oder ein RX Command (MainCmd bzw. `p.packet[0] > 0x80`) ist ?
Das wäre auch für einen Promiscuous mode sinnvoll.

Danach können wir immer noch die Inverter Serial und die Paket Nummer prüfen, aber das gehört m.E. dann bereits zum Paket Zähler, CRC8 Verifikation und anschließenden Payload Analyse.
Für ein möglichst schnelles Umschalten auf Empfang für das nächste RX Paket könnte man das eigentlich auch erst nachgelagert für den kompletten empfangenen Paketstack machen.
Wenn ein Paket fehlt, dann müssen wir eh erst wieder in den TX Modus wechseln.

rejoe2 replied:
Die Prüfung auf den Inverter sollte in jedem Fall hier bleiben, die war nämlich früher nur an einer anderen Stelle. Das hat dann häufiger zu kaputten payloads geführt, weil die Daten eben vom falschen Inverter stammen...

Zum Rest sollte vielleich @lumapu was sagen, da stehe ich grade im Wald, was gemeint sein könnte.

rejoe2 committed:
non blocking - first review

- discord 0.8.5902
- stay longer on expected rx channel to wait for first frame 
- stick to looping over 5 channels for MI (after first frame or timeout)
